File: freeq-android/freeq/src/main/java/com/freeq/ui/components/UserProfileSheet.kt
package com.freeq.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Block
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material.icons.filled.VerifiedUser
import androidx.compose.material.icons.automirrored.filled.OpenInNew
import androidx.compose.material.icons.automirrored.filled.Chat
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.freeq.model.ChatMessage
import com.freeq.model.AppState
import com.freeq.model.AvatarCache
import com.freeq.model.BlueskyProfile
import com.freeq.model.SenderIdentity
import com.freeq.ui.theme.FreeqColors
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun UserProfileSheet(
    nick: String,
    appState: AppState,
    origin: String? = null,
    /** The message this sheet was opened from, when there is one. Its own
     *  tags are evidence: when live identity can't answer, the row does —
     *  the SDK owns that precedence. */
    anchor: ChatMessage? = null,
    onDismiss: () -> Unit,
    onNavigateToDM: (String) -> Unit
) {
    var profile by remember { mutableStateOf<BlueskyProfile?>(null) }
    var loading by remember { mutableStateOf(true) }
    var showReportDialog by remember { mutableStateOf(false) }
    var showIdentityProof by remember { mutableStateOf(false) }
    val uriHandler = LocalUriHandler.current
    val isOwnProfile = nick.equals(appState.nick.value, ignoreCase = true)
    // Identity is the server-bound DID, never the nick. Resolve the DID
    // from our own session (self) or a channel member entry / account-tag
    // map; with no DID there is no Bluesky profile to show (no
    // nick-as-handle guessing).
    val resolvedDid = if (isOwnProfile) {
        appState.authenticatedDID.value
    } else {
        appState.didForNick(nick)
    }
    val isBlocked = appState.isBlocked(nick, resolvedDid)
    // The SDK owns the precedence: live identity first, then the anchoring
    // row's evidence, then the lookup machine. A binding remembered from an
    // earlier session never votes — that hole is how the same absent sender
    // used to read differently here than on web.
    val claim = com.freeq.ffi.claimForSender(
        com.freeq.ffi.MessageClaimInput(
            account = anchor?.account,
            origin = origin,
            senderPresent = appState.isNickPresent(nick),
            senderLiveDid = if (isOwnProfile) appState.authenticatedDID.value else appState.liveDidForNick(nick),
            rowTimeUnix = anchor?.timestamp?.let { (it.time / 1000).toULong() },
        ),
        appState.personLookup(nick),
    )

    LaunchedEffect(nick) {
        // Ask who this is, and let the card say an ask is out — so "unknown"
        // is only ever shown after an answer, never before one.
        if (origin == null && !isOwnProfile) appState.lookUpIdentity(nick)
        profile = withContext(Dispatchers.IO) {
            AvatarCache.fetchProfileIfNeeded(nick, resolvedDid)
        }
        loading = false
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false),
        containerColor = MaterialTheme.colorScheme.background,
        dragHandle = { BottomSheetDefaults.DragHandle() }
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(bottom = 32.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            // Avatar
            UserAvatar(nick = nick, size = 80.dp)

            Spacer(modifier = Modifier.height(16.dp))

            // Nick + verified badge
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp)
            ) {
                Text(
                    text = nick,
                    fontSize = 22.sp,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onBackground
                )
                // The mark follows the identifier, not the avatar fetch, and
                // it is the same rule the message row applies.
                if (claim.showsMark) {
                    Icon(
                        Icons.Default.CheckCircle,
                        contentDescription = "AT Protocol identity — tap for the proof",
                        tint = FreeqColors.accent,
                        modifier = Modifier
                            .size(18.dp)
                            .clickable { showIdentityProof = true }
                    )
                }
            }

            // What this client can honestly say about who this is — the same
            // claim rule and language as every other surface. An ask that is
            // out shows as motion, not as a sentence nobody can finish reading
            // before it is replaced.
            if (claim.isPending) {
                Spacer(modifier = Modifier.height(4.dp))
                CircularProgressIndicator(
                    modifier = Modifier.size(16.dp),
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    strokeWidth = 2.dp
                )
            }
            claim.label?.let { label ->
                Text(
                    text = label,
                    fontSize = 12.sp,
                    color = if (claim.showsMark) FreeqColors.accent
                    else MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            // The proof sheet is where the explanation lives; this card
            // carries the sentence only for someone with no DID, where there
            // is no sheet to open and the card is the whole answer.
            if (claim.did.isNullOrEmpty()) {
                claim.line?.let { line ->
                    Text(
                        text = line,
                        fontSize = 12.sp,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        textAlign = TextAlign.Center,
                        modifier = Modifier.padding(horizontal = 32.dp)
                    )
                }
            }

            // Away status
            appState.awayMessage(nick)?.let { awayMsg ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp)
                ) {
                    Box(
                        modifier = Modifier
                            .size(8.dp)
                            .background(FreeqColors.warning, CircleShape)
                    )
                    Text(
                        text = "Away",
                        fontSize = 12.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = FreeqColors.warning
                    )
                }
                if (awayMsg.isNotEmpty()) {
                    Text(
                        text = awayMsg,
                        fontSize = 13.sp,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }

            // Display name
            profile?.displayName?.takeIf { it.isNotEmpty() }?.let { displayName ->
                Text(
                    text = displayName,
                    fontSize = 15.sp,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            // Handle
            profile?.let { p ->
                Text(
                    text = "@${p.handle}",
                    fontSize = 13.sp,
                    color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f)
                )
            }

            // Bio
            profile?.description?.takeIf { it.isNotEmpty() }?.let { bio ->
                Spacer(modifier = Modifier.height(12.dp))
                Text(
                    text = bio,
                    fontSize = 14.sp,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(horizontal = 32.dp)
                )
            }

            // Stats
            profile?.let { p ->
                Spacer(modifier = Modifier.height(16.dp))
                Row(
                    horizontalArrangement = Arrangement.spacedBy(24.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    StatItem(count = p.followersCount ?: 0, label = "Followers")
                    StatItem(count = p.followsCount ?: 0, label = "Following")
                    StatItem(count = p.postsCount ?: 0, label = "Posts")
                }
            }

            Spacer(modifier = Modifier.height(20.dp))

            // Action buttons
            Column(
                modifier = Modifier.padding(horizontal = 24.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                // Message button (hidden for own profile)
                if (!isOwnProfile) {
                    Button(
                        onClick = {
                            onDismiss()
                            onNavigateToDM(nick)
                        },
                        modifier = Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(10.dp),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = FreeqColors.accent,
                            contentColor = MaterialTheme.colorScheme.onPrimary
                        )
                    ) {
                        Icon(
                            Icons.AutoMirrored.Filled.Chat,
                            contentDescription = null,
                            modifier = Modifier.size(16.dp)
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                        Text("Message", fontWeight = FontWeight.SemiBold)
                    }
                }

                // View on Bluesky button
                profile?.let { p ->
                    Surface(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable {
                                uriHandler.openUri("https://bsky.app/profile/${p.handle}")
                            },
                        shape = RoundedCornerShape(10.dp),
                        color = MaterialTheme.colorScheme.surfaceVariant
                    ) {
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(vertical = 12.dp),
                            horizontalArrangement = Arrangement.Center,
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Icon(
                                Icons.AutoMirrored.Filled.OpenInNew,
                                contentDescription = null,
                                modifier = Modifier.size(16.dp),
                                tint = MaterialTheme.colorScheme.onBackground
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(
                                "View on Bluesky",
                                fontWeight = FontWeight.Medium,
                                color = MaterialTheme.colorScheme.onBackground
                            )
                        }
                    }
                }

                // The identity claim's evidence: this person's DID and the key
                // they sign with. Reachable for everyone we hold a DID for,
                // including the senders who carry no ✓ — a did:key bot and a
                // relayed stranger have a proof view too, and it says what it
                // can support rather than nothing at all.
                if (!resolvedDid.isNullOrEmpty()) {
                    Surface(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable { showIdentityProof = true },
                        shape = RoundedCornerShape(10.dp),
                        color = MaterialTheme.colorScheme.surfaceVariant
                    ) {
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(vertical = 12.dp),
                            horizontalArrangement = Arrangement.Center,
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Icon(
                                Icons.Default.VerifiedUser,
                                contentDescription = null,
                                modifier = Modifier.size(16.dp),
                                tint = MaterialTheme.colorScheme.onBackground
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(
                                "Identity proof",
                                fontWeight = FontWeight.Medium,
                                color = MaterialTheme.colorScheme.onBackground
                            )
                        }
                    }
                }

                // Safety actions (hidden for own profile)
                if (!isOwnProfile) {
                    Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                        OutlinedButton(
                            onClick = { showReportDialog = true },
                            modifier = Modifier.weight(1f),
                            shape = RoundedCornerShape(10.dp),
                            colors = ButtonDefaults.outlinedButtonColors(
                                contentColor = FreeqColors.danger
                            )
                        ) {
                            Icon(
                                Icons.Default.Flag,
                                contentDescription = null,
                                modifier = Modifier.size(16.dp)
                            )
                            Spacer(modifier = Modifier.width(6.dp))
                            Text("Report", fontWeight = FontWeight.Medium)
                        }
                        OutlinedButton(
                            onClick = {
                                if (isBlocked) {
                                    appState.unblockUser(nick, resolvedDid)
                                } else {
                                    appState.blockUser(nick, resolvedDid)
                                    appState.errorMessage.value = "Blocked $nick"
                                }
                            },
                            modifier = Modifier.weight(1f),
                            shape = RoundedCornerShape(10.dp),
                            colors = ButtonDefaults.outlinedButtonColors(
                                contentColor = FreeqColors.danger
                            )
                        ) {
                            Icon(
                                Icons.Default.Block,
                                contentDescription = null,
                                modifier = Modifier.size(16.dp)
                            )
                            Spacer(modifier = Modifier.width(6.dp))
                            Text(
                                if (isBlocked) "Unblock" else "Block",
                                fontWeight = FontWeight.Medium
                            )
                        }
                    }
                }
            }

            // Loading indicator
            if (loading) {
                Spacer(modifier = Modifier.height(16.dp))
                CircularProgressIndicator(
                    modifier = Modifier.size(24.dp),
                    color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f),
                    strokeWidth = 2.dp
                )
            }
        }
    }

    // The proof behind the identity claim this card makes — opened from the
    // card's own mark, over the card, so the person stays the subject.
    if (showIdentityProof) {
        VerifiedProofSheet(
            request = ProofRequest.Identity(
                did = claim.did,
                nick = nick,
                handle = profile?.handle,
                displayName = profile?.displayName?.takeIf { it.isNotEmpty() },
                origin = origin,
                account = anchor?.account,
                rowTimeUnix = anchor?.timestamp?.let { (it.time / 1000).toULong() },
                senderPresent = appState.isNickPresent(nick),
                senderLiveDid = if (isOwnProfile) appState.authenticatedDID.value else appState.liveDidForNick(nick),
            ),
            onDismiss = { showIdentityProof = false },
            appState = appState
        )
    }

    // Report reason picker — on choice: report (log) + block + snackbar
    if (showReportDialog) {
        ReportReasonDialog(
            nick = nick,
            onReport = { reason ->
                appState.reportUser(nick, resolvedDid, reason)
                appState.errorMessage.value = "Reported $nick — user blocked"
                showReportDialog = false
            },
            onDismiss = { showReportDialog = false }
        )
    }
}

@Composable
private fun StatItem(count: Int, label: String) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(
            text = formatCount(count),
            fontSize = 16.sp,
            fontWeight = FontWeight.Bold,
            color = MaterialTheme.colorScheme.onBackground
        )
        Text(
            text = label,
            fontSize = 11.sp,
            color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f)
        )
    }
}

private fun formatCount(n: Int): String {
    if (n >= 1_000_000) return "${n / 1_000_000}M"
    if (n >= 1_000) return "${n / 1_000}K"
    return "$n"
}

File: freeq-macos/freeq-macos/Views/Detail/UserProfileSheet.swift
import SwiftUI

/// Rich user profile sheet: avatar, Bluesky info, shared channels, DID.
struct UserProfileSheet: View {
    @Environment(AppState.self) private var appState
    @Environment(\.dismiss) private var dismiss
    let nick: String
    /// The peer that relayed the message this card was opened from. Anything
    /// relayed is peer-vouched and claims nothing here.
    var origin: String? = nil
    /// Opened from a local message whose sender has no DID on file — already
    /// the guest answer, so no WHOIS is sent and none is needed.
    /// The anchoring row's evidence, when opened from a message.
    var account: String? = nil
    var rowTime: Date? = nil
    var senderPresent: Bool = false

    @State private var reportTarget: ReportTarget?
    @State private var showProof = false

    private var profile: ProfileCache.Profile? {
        ProfileCache.shared.profile(for: nick)
    }
    private var did: String? {
        // ProfileCache only knows AT-resolved profiles; didForNick covers
        // everyone the app has learned a binding for, did:key bots included.
        ProfileCache.shared.did(for: nick) ?? appState.didForNick(nick)
    }
    // The SDK owns the precedence: live identity first, then the anchoring
    // row's evidence, then the lookup machine. A binding remembered from an
    // earlier session never votes — that is how one absent sender used to
    // read differently here than on web.
    private var claim: IdentityClaim {
        claimForSender(
            input: MessageClaimInput(
                account: account,
                origin: origin,
                senderPresent: senderPresent || appState.isNickPresent(nick),
                senderLiveDid: isSelf ? appState.authenticatedDID : appState.liveDidForNick(nick),
                rowTimeUnix: rowTime.map { UInt64($0.timeIntervalSince1970) }
            ),
            lookup: appState.personLookup(for: nick)
        )
    }
    private var isSelf: Bool {
        nick.lowercased() == appState.nick.lowercased()
    }

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack {
                Spacer()
                Button { dismiss() } label: {
                    Image(systemName: "xmark.circle.fill")
                        .font(.title3)
                        .foregroundStyle(.secondary)
                }
                .buttonStyle(.plain)
            }
            .padding(.trailing, 12)
            .padding(.top, 8)

            // Scrollable body — content length varies (bio length, shared-channel
            // count), so it must scroll rather than clip the pinned action buttons.
            ScrollView {
                VStack(spacing: 0) {
                    // Avatar
                    AvatarView(nick: nick, size: 72)
                        .padding(.top, 4)

                    // Name
                    if let displayName = profile?.displayName, !displayName.isEmpty {
                        Text(displayName)
                            .font(.title2.weight(.bold))
                            .padding(.top, 8)
                        Text("@\(profile?.handle ?? nick)")
                            .font(.subheadline)
                            .foregroundStyle(.secondary)
                    } else {
                        Text(nick)
                            .font(.title2.weight(.bold))
                            .padding(.top, 8)
                    }

                    // What this client can honestly say about who this is —
                    // the same claim rule and language as every other surface.
                    // An ask that is out shows as motion, not as a sentence
                    // nobody can finish reading before it is replaced.
                    if claim.isPending {
                        ProgressView().controlSize(.small)
                    }
                    if let label = claim.label {
                        Text(label)
                            .font(.caption)
                            .foregroundStyle(claim.showsMark ? Theme.verified : Theme.textTertiary)
                    }
                    // The proof sheet is where the explanation lives; this
                    // card carries the sentence only for someone with no DID,
                    // where there is no sheet to open and the card is the
                    // whole answer. Everyone else gets the label, and the
                    // words are one click away behind "Identity proof".
                    if claim.did == nil, let line = claim.line {
                        Text(line)
                            .font(.caption)
                            .foregroundStyle(.secondary)
                            .multilineTextAlignment(.center)
                            .fixedSize(horizontal: false, vertical: true)
                            .padding(.horizontal, 24)
                            .padding(.top, 2)
                    }

                    // DID — click for the identity/signing-key proof card
                    if let did {
                        Button { showProof = true } label: {
                            // One explicit stack: two sibling views in a
                            // button's label lay out side by side, which put
                            // the affordance next to the identifier instead of
                            // under it.
                            VStack(spacing: 3) {
                                HStack(spacing: 4) {
                                    if claim.showsMark {
                                        Image(systemName: "checkmark.seal.fill")
                                            .font(.caption)
                                            .foregroundStyle(.blue)
                                    }
                                    Text(did)
                                        .font(.caption2.monospaced())
                                        .foregroundStyle(.secondary)
                                        .lineLimit(1)
                                        .truncationMode(.middle)
                                }
                                // A monospace string doesn't look like a way
                                // in, and for a self-created identity it is
                                // the only one — no mark on a row leads here.
                                HStack(spacing: 2) {
                                    Text("Identity proof")
                                        .font(.caption)
                                    Image(systemName: "chevron.right")
                                        .font(.caption2)
                                }
                                .foregroundStyle(Theme.accent)
                            }
                            // Constrain the row to the card so middle
                            // truncation engages instead of overflowing.
                            .frame(maxWidth: .infinity)
                            .padding(.horizontal, 24)
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                        .padding(.top, 4)
                        .help(claim.label.map { "\($0) — click for proof" } ?? "Click for proof")
                    }

                    // Bio
                    if let bio = profile?.description, !bio.isEmpty {
                        Text(bio)
                            .font(.body)
                            .foregroundStyle(.secondary)
                            .multilineTextAlignment(.center)
                            .padding(.horizontal, 24)
                            .padding(.top, 8)
                    }

                    // Stats
                    if let p = profile {
                        HStack(spacing: 24) {
                            StatItem(label: "Posts", value: p.postsCount ?? 0)
                            StatItem(label: "Followers", value: p.followersCount ?? 0)
                            StatItem(label: "Following", value: p.followsCount ?? 0)
                        }
                        .padding(.top, 12)
                    }

                    Divider().padding(.vertical, 12)

                    // Shared channels
                    let sharedChannels = appState.channels.filter { ch in
                        ch.members.contains(where: { $0.nick.lowercased() == nick.lowercased() })
                    }
                    if !sharedChannels.isEmpty {
                        VStack(alignment: .leading, spacing: 4) {
                            Text("Shared Channels")
                                .font(.caption.weight(.semibold))
                                .foregroundStyle(.secondary)
                                .padding(.horizontal, 24)
                                .frame(maxWidth: .infinity, alignment: .leading)
                            ForEach(sharedChannels) { ch in
                                Button {
                                    appState.activeChannel = ch.name
                                    dismiss()
                                } label: {
                                    HStack {
                                        Image(systemName: "number")
                                            .font(.caption)
                                            .foregroundStyle(.secondary)
                                        Text(ch.name.replacingOccurrences(of: "#", with: ""))
                                            .font(.body)
                                        Spacer()
                                        Text("\(ch.members.count)")
                                            .font(.caption)
                                            .foregroundStyle(.tertiary)
                                    }
                                    .padding(.horizontal, 24)
                                    .padding(.vertical, 4)
                                    .contentShape(Rectangle())
                                }
                                .buttonStyle(.plain)
                            }
                        }
                    }
                }
            }

            // Actions — pinned below the scroll area so they never clip.
            VStack(spacing: 8) {
                HStack(spacing: 12) {
                    Button("Send Message") {
                        appState.openDM(with: nick)
                        dismiss()
                    }
                    .buttonStyle(.borderedProminent)
                    .controlSize(.regular)

                    // Only an AT Protocol identity has a Bluesky page; a
                    // self-issued did:key resolves nowhere.
                    if claim.state == .atProtocol,
                       let handle = profile?.handle ?? did.map({ _ in nick }),
                       let blueSkyURL = Validation.makeBlueSkyProfileURL(handle: handle) {
                        Link("View on Bluesky", destination: blueSkyURL)
                            .font(.body)
                    }
                }

                // Safety
                if !isSelf {
                    HStack(spacing: 16) {
                        if appState.isBlocked(nick: nick, did: did) {
                            Button("Unblock") {
                                appState.unblockUser(nick: nick, did: did)
                            }
                            .buttonStyle(.plain)
                            .font(.caption)
                            .foregroundStyle(Theme.accent)
                        } else {
                            Button("Report…") {
                                reportTarget = ReportTarget(nick: nick, did: did)
                            }
                            .buttonStyle(.plain)
                            .font(.caption)
                            .foregroundStyle(Theme.danger)
                            Button("Block") {
                                appState.blockUser(nick: nick, did: did)
                                dismiss()
                            }
                            .buttonStyle(.plain)
                            .font(.caption)
                            .foregroundStyle(Theme.danger)
                        }
                    }
                }
            }
            .padding(.top, 12)
            .padding(.bottom, 16)
        }
        .frame(width: 340, height: 480)
        .reportDialog($reportTarget) { t, reason in
            appState.reportUser(nick: t.nick, did: t.did, reason: reason)
            dismiss()
        }
        .sheet(isPresented: $showProof) {
            VerifiedProofSheet(
                did: claim.did,
                handle: profile?.handle,
                displayName: profile?.displayName,
                nick: nick,
                origin: origin,
                account: account,
                rowTimeUnix: rowTime.map { UInt64($0.timeIntervalSince1970) },
                senderPresent: senderPresent || appState.isNickPresent(nick)
            )
            .environment(appState)
        }
        .onAppear {
            // Ask who this is, and let the card say an ask is out — the same
            // background WHOIS as before, tracked so "unknown" is only ever
            // shown after an answer, never before one.
            if origin == nil, !isSelf { appState.lookUpIdentity(nick: nick) }
            // Fetch profile if we have DID
            if let did, profile == nil {
                ProfileCache.shared.fetchProfile(nick: nick, did: did)
            }
        }
    }
}

struct StatItem: View {
    let label: String
    let value: Int

    var body: some View {
        VStack(spacing: 2) {
            Text("\(value)")
                .font(.headline)
            Text(label)
                .font(.caption2)
                .foregroundStyle(.secondary)
        }
    }
}

// MARK: - Previews

// The canvas renders these without building or running the app. Bio, stats and
// shared channels stay empty — they come from a live profile fetch — so what
// these show is the identity block and the identifier row beneath it, which is
// where the layout actually varies by state.

#Preview("Profile — AT Protocol") {
    ProfileCache.shared.setDid("did:plc:4qsyxmnsblo4luuycm3572bq", for: "chad")
    return UserProfileSheet(nick: "chad")
        .environment(AppState())
}

#Preview("Profile — self-created") {
    ProfileCache.shared.setDid("did:key:z6MkpdyZFX7tagEUaCBp8bwVgqBqzvLEDdi8E", for: "lobot")
    return UserProfileSheet(nick: "lobot")
        .environment(AppState())
}

#Preview("Profile — relayed") {
    ProfileCache.shared.setDid("did:plc:padwfc6z5dke5g7c3nzratdy", for: "nandi-test")
    return UserProfileSheet(nick: "nandi-test", origin: "irc.freeq.at")
        .environment(AppState())
}

#Preview("Profile — nobody on file") {
    UserProfileSheet(nick: "stranger")
        .environment(AppState())
}
